File: shared-packages/packages/rubot-logger-ts/src/logger.ts
import type { LogLevel, LogEnvelope, ErrorDetail } from "./envelope";
import { getConfig } from "./config";
import { getContext, type RequestContext } from "./context";

export interface LogOptions {
  error?: Error;
  extra?: Record<string, unknown>;
  agent?: Record<string, unknown>;
}

export class RubotLogger {
  private component: string;
  private contextKey: object | null;

  constructor(component: string, contextKey?: object) {
    this.component = component;
    this.contextKey = contextKey ?? null;
  }

  withContext(key: object): RubotLogger {
    return new RubotLogger(this.component, key);
  }

  private emit(
    level: LogLevel,
    eventType: string,
    message: string,
    opts?: LogOptions,
  ): void {
    const cfg = getConfig();
    const ctx: RequestContext | null = this.contextKey
      ? getContext(this.contextKey)
      : null;

    let errDetail: ErrorDetail | undefined;
    if (opts?.error) {
      errDetail = {
        type: opts.error.constructor.name,
        message: opts.error.message,
        stack: opts.error.stack,
      };
    }

    const envelope: LogEnvelope = {
      timestamp: new Date().toISOString(),
      log_level: level,
      service: cfg.service,
      component: this.component,
      environment: cfg.environment,
      deployment_hash: cfg.deploymentHash,
      tenant_id: ctx?.tenantId ?? null,
      sender_id: ctx?.senderId ?? null,
      chat_source_session_id: ctx?.chatSourceSessionId ?? null,
      trace_id: ctx?.traceId ?? null,
      event_type: eventType,
      message,
      error: errDetail ?? null,
      extra: opts?.extra ?? {},
      agent: opts?.agent ?? null,
    };

    const clean = Object.fromEntries(
      Object.entries(envelope).filter(([, v]) => v !== null && v !== undefined),
    );
    const line = JSON.stringify(clean);

    switch (level) {
      case "ERROR":
        console.error(line);
        break;
      case "WARN":
        console.warn(line);
        break;
      case "DEBUG":
        console.debug(line);
        break;
      default:
        console.log(line);
        break;
    }
  }

  debug(eventType: string, message: string, opts?: LogOptions): void {
    this.emit("DEBUG", eventType, message, opts);
  }

  info(eventType: string, message: string, opts?: LogOptions): void {
    this.emit("INFO", eventType, message, opts);
  }

  warn(eventType: string, message: string, opts?: LogOptions): void {
    this.emit("WARN", eventType, message, opts);
  }

  error(eventType: string, message: string, opts?: LogOptions): void {
    this.emit("ERROR", eventType, message, opts);
  }
}

export function getLogger(component: string): RubotLogger {
  return new RubotLogger(component);
}

File: shared-packages/packages/rubot-logger/rubot_logger/logger.py
from __future__ import annotations

import logging
import traceback
from typing import Any

from rubot_logger.config import get_config
from rubot_logger.context import (
    chat_source_session_id_var,
    sender_id_var,
    tenant_id_var,
    trace_id_var,
)
from rubot_logger.envelope import ErrorDetail, LogEnvelope

_LEVEL_MAP = {
    "DEBUG": logging.DEBUG,
    "INFO": logging.INFO,
    "WARN": logging.WARNING,
    "ERROR": logging.ERROR,
}


class RubotLogger:
    def __init__(self, component: str) -> None:
        self._component = component
        self._stdlib = logging.getLogger(component)

    def _emit(
        self,
        level: str,
        event_type: str,
        message: str,
        *,
        error: Exception | None = None,
        agent: dict[str, Any] | None = None,
        extra: dict[str, Any] | None = None,
    ) -> None:
        cfg = get_config()
        err_detail = None
        if error is not None:
            err_detail = ErrorDetail(
                type=type(error).__name__,
                message=str(error),
                stack=traceback.format_exc(),
            )

        envelope = LogEnvelope(
            log_level=level,
            service=cfg.service,
            component=self._component,
            environment=cfg.environment,
            deployment_hash=cfg.deployment_hash,
            tenant_id=tenant_id_var.get(),
            sender_id=sender_id_var.get(),
            chat_source_session_id=chat_source_session_id_var.get(),
            trace_id=trace_id_var.get(),
            event_type=event_type,
            message=message,
            error=err_detail,
            agent=agent,
            extra=extra or {},
        )

        self._stdlib.log(
            _LEVEL_MAP.get(level, logging.INFO),
            envelope.model_dump_json(exclude_none=True),
        )

    def debug(self, event_type: str, message: str, **kwargs: Any) -> None:
        self._emit("DEBUG", event_type, message, **kwargs)

    def info(self, event_type: str, message: str, **kwargs: Any) -> None:
        self._emit("INFO", event_type, message, **kwargs)

    def warn(self, event_type: str, message: str, **kwargs: Any) -> None:
        self._emit("WARN", event_type, message, **kwargs)

    def error(self, event_type: str, message: str, **kwargs: Any) -> None:
        self._emit("ERROR", event_type, message, **kwargs)


def get_logger(component: str) -> RubotLogger:
    return RubotLogger(component)
